File: src/main/java/com/matag/cards/sets/MtgSets.java
package com.matag.cards.sets;

import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.core.io.Resource;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.matag.cards.ResourceLoader;

@Component
public class MtgSets {
    private final Map<String, MtgSet> SETS = new LinkedHashMap<>();

    public MtgSets(ObjectMapper objectMapper, ResourceLoader resourceLoader) {
        Resource[] setResources = resourceLoader.getSetsFileNames();
        for (Resource setResource : setResources) {
            try {
                MtgSet mtgSet = objectMapper.readValue(setResource.getInputStream(), MtgSet.class);
                SETS.put(mtgSet.getCode(), mtgSet);

            } catch (Exception e) {
                throw new RuntimeException("Failed to load set: " + setResource.getDescription(), e);
            }
        }
    }

    public Map<String, MtgSet> getSets() {
        return SETS;
    }

    public MtgSet getSet(String code) {
        return SETS.get(code);
    }
}

File: src/main/kotlin/com/matag/cards/sets/MtgSets.kt
package com.matag.cards.sets

import com.fasterxml.jackson.databind.ObjectMapper
import com.matag.cards.ResourceLoader
import org.springframework.core.io.Resource
import org.springframework.stereotype.Component

@Component
class MtgSets(val objectMapper: ObjectMapper, resourceLoader: ResourceLoader) {
    val sets: Map<String, MtgSet> = resourceLoader.getSetsFileNames()
        .map { parseSet(it) }
        .associateBy { it.code }

    fun get(code: String) =
        sets[code] ?: throw Exception("Set $code not found!")

    private fun parseSet(resource: Resource): MtgSet =
        try {
            objectMapper.readValue(resource.inputStream, MtgSet::class.java)
        } catch (e: Exception) {
            throw Exception("Failed to load set: " + resource.description, e)
        }
}
